include/ginkgo/core/base/polymorphic_object.hpp

```diff
@@ -504,10 +544,11 @@ std::unique_ptr<R, std::function<void(R*)>> copy_and_convert_to_impl(
 {
     auto obj_as_r = dynamic_cast<R*>(obj);
     if (obj_as_r != nullptr && obj->get_executor() == exec) {
+        // FIXME: this breaks lifetimes
         return {obj_as_r, [](R*) {}};
```

Comment:
It returns a unique_ptr, which may or may not delete the object. Thus the user has no way of knowing if this will respect the lifetime of the object being passed in.

Reply:
we should probably move it to the core header not public header.
from raw_pointer, I do not think we can safely maintain the lifetime.